### PR TITLE
Device discovery

### PR DESCRIPTION
**Pull request by Foxinatel**
This PR implements the following network flow diagram: networking flow diagram

Allowing GGRS sessions to be created over a local network using multicast-based device discovery

### CODE REVIEW

**Comment by Foxinatel** [comment]
A few features are missing from this PR:
- Being able to select at runtime whether you want to host or join. Ideally this will be done with a UI, which needs some further consideration
- Anchor sharing, as this relies on foundational code from the `scene-v2` branch and also doing developer verification through Meta

**Comment by Foxinatel** [comment]
I'm going to temporarily close this PR for now. It may be re-opened in future.
While this branch does work, we've decided that for the purposes of our tech demo we want to just have a single-player game
Also, this branch isn't particularly useful until we can figure out how to create interactable menus so the player can select whether they want to host or join

**Comment by rosefromthedead** [comment]
reopening to keep it in the tracker, i'll mark as draft if i can figure out how

**Comment by Foxinatel** [comment]
This branch is nearly ready for review now. Before reopening this commit I would like the following PRs to land, as this code depends on both of these:
- #46
- https://github.com/goudanough/bevy_oxr/pull/5

One extra thing to consider as well is that currently we panic immediately if we're on a network that doesn't let us establish a multicast. We should probably add some visual indicators or something explaining what's gone wrong.

File: src/network.rs [hunk 1]
```diff
@@ -36,11 +79,16 @@ pub struct PlayerLeftPalm;
 #[derive(Component)]
 pub struct PlayerRightPalm;
 
+// Why this number? I asked a random number generator!
+const GGRS_PORT: u16 = 47511;
```
Comment (AmosHolland) [inline src/network.rs:83]: based.

File: src/network.rs [hunk 2]
```diff
@@ -49,95 +97,340 @@ impl Plugin for NetworkPlugin {
...
+) {
+    let emit = &emit.0;
+
+    // First we do a listen to see if we've got any incoming connections
+    if let Some((stream, _)) = emit.accept() {
+        commands.insert_resource(HostConnection(stream));
+        state.0 = Some(NetworkingState::ClientWaitForData);
+    } else {
+        // If we there are no incoming requests then we emit a new multicast message
+        // TODO: put this on a timer
```
Comment (AmosHolland) [inline src/network.rs:326]: If adding a timer wouldn't take very long should we do it now? I can imagine this getting merged, and then after that we never look at it again so we never fix this on a timer and we're just sending out multicast messages during every frame of setup forever.

**Review by AmosHolland — CHANGES_REQUESTED** [review]
Overall looks good, really good annotations as well, I think I could actually follow what was happening. 

Should we move network.rs into the network folder, and just have it as `mod.rs` in there? Just seems odd to have a network.rs and a network folder.

File: src/network.rs (hunk 2 above)
Comment (Foxinatel) [reply src/network.rs]: Fixed with https://github.com/goudanough/WizARds/pull/15/commits/5a01b76ac60e37e428b6b8858e3ca60380e23157

**Review by AmosHolland — APPROVED** [review]
Looks good, nice work!

File: src/network/mod.rs [hunk 3]
```diff
@@ -339,6 +340,7 @@ fn client_establish_tcp(
     } else if timer.0.finished() {
         emit.emit();
     }
+    timer.0.tick(time.delta());
```
Comment (AmosHolland) [inline src/network/mod.rs:343]: If we're only checking the timer in once place I think the best way to do this is
```rust
if (...) { ...
} else if timer.0.tick(time.delta()).just_finished() { ... }

**Review by AmosHolland — CHANGES_REQUESTED** [review]
looks good, minor change to ticking the timer.

**Comment by rosefromthedead** [comment]
Please undo the Cargo.lock changes (other than the bevy_oxr update ofc). There's a lot of speech changes in here, are they supposed to be? Especially the addition of grammar = ["fireball", "lightning"], I think some other changes have made their way in by accident

**Comment by Foxinatel** [comment]
Yeah I think the best course of action for this PR is just to close it, because there were some half-baked stuff that got changed and/or reverted in `anchor-sync`

**Pull request closed without merging**